### PR TITLE
DiscIO: Use std::optional for GetTitleID instead of pointer

### PR DESCRIPTION
**Pull request by leoetlino**
This makes the interface slightly cleaner and a bit more consistent with the other getters. Still not fully the same, since the others don't really handle failures with std::optional; but at least the value is returned by value now, as opposed to having the function take a pointer to a u64.

### CODE REVIEW

File: Source/Core/DolphinQt2/GameList/GameFile.cpp [hunk 1]
```diff
@@ -161,7 +161,8 @@ bool GameFile::TryLoadVolume()
 
   m_game_id = QString::fromStdString(volume->GetGameID());
   std::string maker_id = volume->GetMakerID();
-  volume->GetTitleID(&m_title_id);
+  if (std::optional<u64> title_id = volume->GetTitleID())
+    m_title_id = *title_id;
```
Comment (iwubcode) [inline Source/Core/DolphinQt2/GameList/GameFile.cpp]: Is there a reason we don't make m_title_id a std optional as well?

File: Source/Core/Core/HW/DVD/DVDThread.cpp [hunk 2]
```diff
@@ -219,11 +219,11 @@ bool UpdateRunningGameMetadata(const DiscIO::Partition& partition, std::optional
 
   if (title_id)
   {
-    u64 volume_title_id;
-    if (!s_disc->GetTitleID(&volume_title_id, partition))
+    const std::optional<u64> volume_title_id = s_disc->GetTitleID(partition);
+    if (!volume_title_id)
```
Comment (JosJuice) [inline Source/Core/Core/HW/DVD/DVDThread.cpp]: It would probably be appropriate to shorten this to `if (!volume_title_id || *volume_title_id != *title_id)`, since `!volume_title_id` is a simple check compared to `!s_disc->GetTitleID(&volume_title_id, partition)`.

File: Source/Core/DolphinQt2/GameList/GameFile.cpp (hunk 1 above)
Comment (JosJuice) [inline Source/Core/DolphinQt2/GameList/GameFile.cpp]: We should probably do that, though it might be better to leave it for a separate PR.

File: Source/Core/Core/HW/DVD/DVDThread.cpp (hunk 2 above)
Comment (leoetlino) [inline Source/Core/Core/HW/DVD/DVDThread.cpp]: Done.

File: Source/Core/DolphinQt2/GameList/GameFile.cpp (hunk 1 above)
Comment (leoetlino) [inline Source/Core/DolphinQt2/GameList/GameFile.cpp]: It looks like the current UI code does not really care about GetTitleID failing and using an uninitialised/zero u64 for the title ID. With a std::optional<u64>, more explicit checks would have to be added though, so I think this would be better left to a future PR.

**Comment by RisingFog** [comment]
Since this was merged, Action Replay/Gecko Codes don't work

https://bugs.dolphin-emu.org/issues/10320
